ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,323 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::counters::{IndexerTableInfoStep, DURATION_IN_SECS, SERVICE_TYPE, TOTAL_SIZE_IN_BYTES};
+use anyhow::Error;
+use aptos_api::context::Context;
+use aptos_api_types::TransactionOnChainData;
+use aptos_logger::{debug, error, info, sample, sample::SampleRate};
+use aptos_sdk::bcs;
+use aptos_storage_interface::{DbReaderWriter, DbWriter};
+use aptos_types::write_set::WriteSet;
+use std::{sync::Arc, time::Duration};
+use tonic::Status;
+
+type EndVersion = u64;
+const LEDGER_VERSION_RETRY_TIME_MILLIS: u64 = 10;
+const DEFAULT_NUM_RETRIES: usize = 3;
+
+pub struct TableInfoParser {
+    pub current_version: u64,
+    pub parser_task_count: u16,
+    pub parser_batch_size: u16,
+    pub context: Arc<Context>,
+}
+
+pub struct TransactionBatchInfo {
+    pub start_version: u64,
+    pub num_transactions_to_fetch: u16,
+}
+
+impl TableInfoParser {
+    pub fn new(
+        context: Arc<Context>,
+        request_start_version: u64,
+        parser_task_count: u16,
+        parser_batch_size: u16,
+    ) -> Self {
+        Self {
+            current_version: request_start_version,
+            parser_task_count,
+            parser_batch_size,
+            context,
+        }
+    }
+
+    /// Fans out a bunch of threads and processes write sets from transactions in parallel.
+    /// Pushes results in parallel to the stream, but only return that the batch is
+    /// fully completed if every job in the batch is successful
+    /// Processing transactions in 2 stages:
+    /// 1. Fetch transactions from storage
+    /// 2. Get write sets from transactions and parse write sets to get handle -> key,value type mapping, write the mapping to the rocksdb
+    pub async fn process_next_batch(
+        &mut self,
+        db: DbReaderWriter,
+    ) -> Vec<Result<EndVersion, Status>> {
+        let mut tasks = vec![];
+        let (batches, total_txns_to_process, ledger_version) = self.get_batches().await;
+        let db_writer = db.writer.clone();
+
+        for batch in batches {
+            let start_time = std::time::Instant::now();
+            let db_writer = db_writer.clone();
+            let context = self.context.clone();
+            let task = tokio::spawn(async move {
+                let raw_txns =
+                    Self::fetch_raw_txns_with_retries(context.clone(), ledger_version, batch).await;
+                Self::parse_table_info(context.clone(), raw_txns.clone(), db_writer.clone(), false)
+                    .expect("[Table Info] Failed to parse table info");
+
+                DURATION_IN_SECS
+                    .with_label_values(&[
+                        SERVICE_TYPE,
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_step(),
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_label(),
+                    ])
+                    .set(start_time.elapsed().as_secs_f64());
```

Comment:
can you rebase on main and follow this new format for logging? It'll be a lot cleaner. https://github.com/aptos-labs/aptos-core/pull/11306/files